dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/SubexpressionPeriodOffsetUtils.java

```diff
@@ -0,0 +1,188 @@
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.analytics.data;
+
+import static org.hisp.dhis.analytics.util.AnalyticsSqlUtils.quote;
+import static org.hisp.dhis.analytics.util.PeriodOffsetUtils.shiftPeriod;
+import static org.hisp.dhis.common.DimensionalObject.DATA_X_DIM_ID;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.hisp.dhis.analytics.DataQueryParams;
+import org.hisp.dhis.common.BaseDimensionalObject;
+import org.hisp.dhis.common.DimensionType;
+import org.hisp.dhis.common.DimensionalItemObject;
+import org.hisp.dhis.common.DimensionalObject;
+import org.hisp.dhis.period.Period;
+
+/**
+ * Utility methods for generating a subExpression query containing periodOffsets.
+ *
+ * <p>These methods help in the translation between report periods (the periods reported to the
+ * user) and data periods (the database periods in which data is fetched). For example a
+ * periodOffset of -1 for a reporting period of '202310' (Oct 2023) means that the data comes from
+ * data period '202309' (Sep 2023).
+ *
+ * @author Jim Grace
+ */
+public class SubexpressionPeriodOffsetUtils {
+  private SubexpressionPeriodOffsetUtils() {
+    throw new UnsupportedOperationException("util");
+  }
+
+  protected static final String SHIFT = "shift";
+
+  protected static final String DELTA = quote("delta");
+
+  protected static final String REPORTPERIOD = quote("reportperiod");
+
+  protected static final String DATAPERIOD = quote("dataperiod");
+
+  /**
+   * For {@link DataQueryParams} containing a subexpression with periodOffsets, joins an inline
+   * value table that maps the reporting periods to data periods, for each distinct periodOffset
+   * value.
+   *
+   * <p>For example, for periods '202309' and '202310' and periods offsets -1 and 0, this would
+   * generate:
+   *
+   * <pre>
+   *       join (values(-1,'202309','202308'),(-1,'202310','202309'),
+   *                   (0,'202309','202309'),(0,'202310','202310'))
+   *       as shift (delta, reportperiod, dataperiod) on dataperiod = "monthly"
+   * </pre>
+   *
+   * @param params parameters with reporting periods
+   * @return a join clause to an inline table to map reporting periods to data periods
+   */
+  protected static String joinPeriodOffsetValues(DataQueryParams params) {
+    List<Period> reportPeriods = getReportPeriods(params);
+    List<Integer> periodOffsets = getPeriodOffsets(params);
+
+    StringBuilder sb = new StringBuilder(" join (values");
+
+    for (Integer delta : periodOffsets) {
+      for (Period reportPeriod : reportPeriods) {
+        Period dataPeriod = shiftPeriod(reportPeriod, delta);
+        sb.append("(")
+            .append(delta)
+            .append(",'")
+            .append(reportPeriod.getIsoDate())
+            .append("','")
+            .append(dataPeriod.getIsoDate())
+            .append("'),");
```

Comment:
This is much less readable then using a template and inserting the values. Is this really performance critical to build a SQL query so we would need to use a `StringBuilder`?

Reply:
Good suggestion, thanks. Performance is not as important here as readability. Changed.